### PR TITLE
Fix LocalDynamicFiltersCollector#getPredicate for multiple co-located joins

### PR DESCRIPTION
**Pull request by raunaqmorarka**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java [hunk 1]
```diff
@@ -26,20 +32,32 @@
...
-    private TupleDomain<Symbol> predicate;
-
-    public LocalDynamicFiltersCollector()
-    {
-        this.predicate = TupleDomain.all();
-    }
+    private Map<Symbol, Domain> dynamicFilterDomainsResult = new HashMap<>();
 
-    public synchronized TupleDomain<Symbol> getPredicate()
+    public synchronized TupleDomain<Symbol> getPredicate(Collection<Symbol> scanAssignments)
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: yes, should have been `probeSymbols`, renamed it now

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java [hunk 2]
```diff
@@ -26,20 +32,32 @@
...
 
-    public synchronized void intersect(TupleDomain<Symbol> predicate)
+    // Domains for different subset of dynamic filters maybe requested by different table scans,
+    // therefore we record and intersect Domain of each dynamic filter separately instead of in a combined TupleDomain
+    public synchronized void intersect(Map<Symbol, Domain> dynamicFilterDomains)
     {
-        this.predicate = this.predicate.intersect(predicate);
+        for (Map.Entry<Symbol, Domain> entry : dynamicFilterDomains.entrySet()) {
+            Domain intersectionDomain = dynamicFilterDomainsResult.get(entry.getKey());
+            if (intersectionDomain == null) {
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: done, thanks

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 3]
```diff
@@ -82,29 +89,43 @@ private synchronized void addPartition(TupleDomain<String> tupleDomain)
...
         }
-        // Convert the predicate to use probe symbols (instead dynamic filter IDs).
-        // Note that in case of a probe-side union, a single dynamic filter may match multiple probe symbols.
+
         ImmutableMap.Builder<Symbol, Domain> builder = ImmutableMap.builder();
-        for (Map.Entry<String, Domain> entry : result.getDomains().get().entrySet()) {
-            Domain domain = entry.getValue();
+        Map<String, Domain> domains = result.getDomains().get();
+        for (Map.Entry<String, Type> entry : filterTypes.entrySet()) {
+            Domain domain = domains.get(entry.getKey());
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: done

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 4]
```diff
@@ -52,20 +53,26 @@
     // Mapping from dynamic filter ID to its build channel indices.
     private final Map<String, Integer> buildChannels;
 
-    private final SettableFuture<TupleDomain<Symbol>> resultFuture;
+    // Mapping from dynamic filter ID to its build channel type.
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: I don't think you need to modify `LocalDynamicFilter` (`TupleDomain` works here fine)

I think you only need to fix `LocalDynamicFiltersCollector`

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java [hunk 5]
```diff
@@ -26,20 +32,26 @@
...
-    private TupleDomain<Symbol> predicate;
-
-    public LocalDynamicFiltersCollector()
-    {
-        this.predicate = TupleDomain.all();
-    }
+    private Map<Symbol, Domain> dynamicFilterDomainsResult = new HashMap<>();
 
-    public synchronized TupleDomain<Symbol> getPredicate()
+    public synchronized TupleDomain<Symbol> getPredicate(Collection<Symbol> probeSymbols)
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: Make `probeSymbols` a `Set` to explicitly avoid duplicates

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java (hunk 4 above)
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: `LocalDynamicFilter` is created per join node. Therefore `TupleDomain` should be fine here (empty domain means join filters everything)

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java (hunk 5 above)
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: done

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 6]
```diff
@@ -21,6 +21,7 @@
 import io.airlift.log.Logger;
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: shortened it

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java (hunk 4 above)
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: Simplified to make changes only in `LocalDynamicFiltersCollector`

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java [hunk 7]
```diff
@@ -26,20 +33,37 @@
      * (e.g. in case of co-located joins).
      */
     @GuardedBy("this")
-    private TupleDomain<Symbol> predicate;
-
-    public LocalDynamicFiltersCollector()
-    {
-        this.predicate = TupleDomain.all();
-    }
+    private Optional<Map<Symbol, Domain>> dynamicFilterDomainsResult = Optional.of(new HashMap<>());
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: Don't need optional here, use empty map initially.

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java [hunk 8]
```diff
@@ -26,20 +33,37 @@
...
+        for (Symbol symbol : probeSymbols) {
+            Domain domain = domains.get(symbol);
+            if (domain != null) {
+                builder.put(symbol, domain);
+            }
+        }
+        return TupleDomain.withColumnDomains(builder.build());
     }
 
+    // Domains for different subset of dynamic filters maybe requested by different table scans,
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: remove this comment

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java [hunk 9]
```diff
@@ -26,20 +33,37 @@
...
+    // therefore we record and intersect Domain of each dynamic filter separately instead of in a combined TupleDomain
     public synchronized void intersect(TupleDomain<Symbol> predicate)
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: done

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 10]
```diff
@@ -86,10 +89,17 @@ private synchronized void addPartition(TupleDomain<String> tupleDomain)
         }
     }
 
-    private TupleDomain<Symbol> convertTupleDomain(TupleDomain<String> result)
+    private Map<Symbol, Domain> convertTupleDomain(TupleDomain<String> result)
     {
         if (result.isNone()) {
-            return TupleDomain.none();
+            ImmutableMap.Builder<Symbol, Domain> builder = ImmutableMap.builder();
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: you could use stream here:
```
buildChannels.keySet().stream()
  .flatMap(filterId -> probeSymbols.get(filderId).stream())
  .collect(toImmutableMap(probeSymbol -> probeSymbol, omain.none(types.get(probeSymbol)));
```

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 11]
```diff
@@ -86,10 +89,17 @@ private synchronized void addPartition(TupleDomain<String> tupleDomain)
...
+            for (String filterId : buildChannels.keySet()) {
+                // Store all matching symbols for each build channel index.
+                for (Symbol probeSymbol : probeSymbols.get(filterId)) {
+                    builder.put(probeSymbol, Domain.none(types.get(probeSymbol)));
+                }
+            }
+            return builder.build();
         }
         // Convert the predicate to use probe symbols (instead dynamic filter IDs).
         // Note that in case of a probe-side union, a single dynamic filter may match multiple probe symbols.
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: use stream here also.

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 12]
```diff
@@ -86,10 +89,17 @@ private synchronized void addPartition(TupleDomain<String> tupleDomain)
         }
     }
 
-    private TupleDomain<Symbol> convertTupleDomain(TupleDomain<String> result)
+    private Map<Symbol, Domain> convertTupleDomain(TupleDomain<String> result)
     {
         if (result.isNone()) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: you could add a similar comment:
```
one of the join build symbols has no non-null values, therefore no symbols can match predicate
```

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java [hunk 13]
```diff
@@ -26,20 +32,24 @@
...
-    {
-        this.predicate = TupleDomain.all();
-    }
+    private Map<Symbol, Domain> dynamicFilterDomainsResult = new HashMap<>();
 
-    public synchronized TupleDomain<Symbol> getPredicate()
+    public synchronized TupleDomain<Symbol> getDynamicFilter(Set<Symbol> probeSymbols)
     {
-        return predicate;
+        ImmutableMap.Builder<Symbol, Domain> builder = ImmutableMap.builder();
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: use streams here

File: presto-main/src/test/java/io/prestosql/sql/planner/TestLocalDynamicFiltersCollector.java [hunk 14]
```diff
@@ -29,25 +33,64 @@
...
+                TupleDomain.withColumnDomains(ImmutableMap.of(
+                        symbol1, Domain.multipleValues(BIGINT, ImmutableList.of(1L, 2L)),
+                        symbol2, Domain.multipleValues(BIGINT, ImmutableList.of(2L, 3L)))));
+
+        collector.addDynamicFilter(toDomainMap(symbol1, 0L));
+        assertEquals(collector.getDynamicFilter(probeSymbols1), TupleDomain.none());
+        assertEquals(collector.getDynamicFilter(probeSymbols2), tupleDomain(symbol2, 2L, 3L));
+        assertEquals(collector.getDynamicFilter(ImmutableSet.of(symbol1, symbol2)), TupleDomain.none());
+
+        collector.addDynamicFilter(ImmutableMap.of(symbol2, Domain.none(BIGINT)));
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestLocalDynamicFiltersCollector.java]: Please also make sure that adding `none` domain for particular symbol doesn't affect other symbols (set them to none domain too)

File: presto-main/src/test/java/io/prestosql/sql/planner/TestLocalDynamicFilter.java [hunk 15]
```diff
@@ -128,17 +132,18 @@ public void testNone()
...
                 ImmutableMap.of("123", 0),
+                TypeProvider.copyOf(ImmutableMap.of(new Symbol("a"), INTEGER)),
                 1);
         assertEquals(filter.getBuildChannels(), ImmutableMap.of("123", 0));
         Consumer<TupleDomain<String>> consumer = filter.getTupleDomainConsumer();
-        ListenableFuture<TupleDomain<Symbol>> result = filter.getResultFuture();
+        ListenableFuture<Map<Symbol, Domain>> result = filter.getResultFuture();
 
         assertFalse(result.isDone());
         consumer.accept(TupleDomain.withColumnDomains(ImmutableMap.of(
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestLocalDynamicFilter.java]: change it to `consumer.accept(TupleDomain.none())`

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java (hunk 12 above)
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: added

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java (hunk 11 above)
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: done

File: presto-main/src/test/java/io/prestosql/sql/planner/TestLocalDynamicFilter.java (hunk 15 above)
Comment (raunaqmorarka) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestLocalDynamicFilter.java]: done

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java (hunk 13 above)
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java]: done

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java (hunk 10 above)
Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: done

File: presto-main/src/test/java/io/prestosql/sql/planner/TestLocalDynamicFiltersCollector.java (hunk 14 above)
Comment (raunaqmorarka) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestLocalDynamicFiltersCollector.java]: Added another test case `TestLocalDynamicFiltersCollector#testCollectorMultipleScansNone`

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java [hunk 16]
```diff
@@ -86,25 +90,23 @@ private synchronized void addPartition(TupleDomain<String> tupleDomain)
...
-        for (Map.Entry<String, Domain> entry : result.getDomains().get().entrySet()) {
-            Domain domain = entry.getValue();
-            // Store all matching symbols for each build channel index.
-            for (Symbol probeSymbol : probeSymbols.get(entry.getKey())) {
-                builder.put(probeSymbol, domain);
-            }
-        }
-        return TupleDomain.withColumnDomains(builder.build());
+        return result.getDomains().get().entrySet().stream()
+                .flatMap(entry -> probeSymbols.get(entry.getKey()).stream()
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: I think loop was better here, please revert. Sorry

Comment (raunaqmorarka) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java]: np, reverted

**Pull request merged**